tensorboard/manager.py

```diff
@@ -292,3 +294,129 @@ def get_all():
+    ),
+)
+
+# Indicates that a call to `start` launched a TensorBoard process, but
+# that process neither exited nor wrote its info file within the allowed
+# timeout period. The process may still be running under the included
+# PID.
+StartTimedOut = collections.namedtuple("StartTimedOut", ("pid",))
+
+
+def start(arguments, timeout=datetime.timedelta(seconds=10)):
+  """Start a new TensorBoard instance, or reuse a compatible one.
+
+  If the cache key determined by the provided arguments and the current
+  working directory (see `cache_key`) matches the cache key of a running
+  TensorBoard process (see `get_all`), that process will be reused.
+
+  Otherwise, a new TensorBoard process will be spawned with the provided
+  arguments, using the `tensorboard` binary from the system path.
+
+  Args:
+    arguments: List of strings to be passed as arguments to
+      `tensorboard`. (If you have a raw command-line string, see
+      `shlex.split`.)
+    timeout: `datetime.timedelta` object describing how long to wait
+      after launching the subprocess before giving up on it. If the
+      subprocess does not write its info file within this time period,
+      it will be left running in the background.
+
+  Returns:
+    A `StartReused`, `StartLaunched`, `StartFailed`, or `StartTimedOut`
+    object.
+  """
+  match = _find_matching_instance(
+      cache_key(
+          working_directory=os.getcwd(),
+          arguments=arguments,
+          configure_kwargs={},
+      ),
+  )
+  if match:
+    return StartReused(info=match)
+
+  (stdout_fd, stdout_path) = tempfile.mkstemp(prefix=".tensorboard-stdout-")
+  (stderr_fd, stderr_path) = tempfile.mkstemp(prefix=".tensorboard-stderr-")
+  start_time = datetime.datetime.now()
+  try:
+    p = subprocess.Popen(
+        ["tensorboard"] + arguments,
+        stdout=stdout_fd,
+        stderr=stderr_fd,
+    )
+  finally:
+    os.close(stdout_fd)
+    os.close(stderr_fd)
+
+  poll_interval_seconds = 0.5
+  end_time = start_time + timeout
+  while datetime.datetime.now() < end_time:
+    time.sleep(poll_interval_seconds)
+    subprocess_result = p.poll()
+    if subprocess_result is not None:
+      return StartFailed(
+          exit_code=subprocess_result,
+          stdout=_maybe_read_file(stdout_path),
+          stderr=_maybe_read_file(stderr_path),
+      )
+    for info in get_all():
+      if info.pid == p.pid and info.start_time >= start_time:
+        return StartLaunched(info=info)
+  else:
+    return StartTimedOut(pid=p.pid)
+
+
+def _find_matching_instance(cache_key):
+  """Find a running TensorBoard instance compatible with the cache key.
+
+  Returns:
+    A `TensorboardInfo` object, or `None` if none matches the cache key.
+  """
+  infos = get_all()
+  candidates = [info for info in infos if info.cache_key == cache_key]
+  for candidate in sorted(candidates, key=lambda x: x.port):
+    # TODO(@wchargin): Check here that the provided port is still live.
```

Comment:
No action required: what do you mean here? The port can be occupied by other program. Do you mean check whether TensorBoard is alive?

Reply:
Yes, check whether TensorBoard is alive. If the TensorBoard process is
killed ungracefully (e.g., with SIGKILL or SIGQUIT), then it won’t get a
chance to clean up its info file, so the files in this directory _may_
be out of date. We can easily check whether the server is alive by just
sending an HTTP request to its `/data/logdir`.